### PR TITLE
compare.cc (default validator): improve float WA message

### PR DESCRIPTION
**Pull request by mpsijm**
The WA message showed `-1` for undefined tolerance values, which was kind of confusing.
Additionally, it was not always clear whether it was the absolute and/or relative tolerance that was exceeded.
This PR adds some extra branches to show a more precise WA message.

The main reason for this change is that the output of this validator is not only visible to jury members, but also to participants for the sample cases, which caused some confusion. But also for jury members, the validator output should be more clear now :slightly_smiling_face: 

I couldn't find any tests for this default validator, but if there are any, I'll be happy to still add some more, to cover all the new branches.
For now, I've tested this validator manually, using some problems from [BAPC 2022](https://2022.bapc.eu/). Problem Failing Flagship has only an absolute tolerance, and problem Imperfect Imperial Units has only a relative tolerance.

@RagnarGrootKoerkamp you'll probably also like this :smile:

### CODE REVIEW

File: sql/files/defaultdata/compare/compare.cc [hunk 1]
```diff
@@ -176,16 +153,46 @@ int main(int argc, char **argv) {
 			if (!isfloat(team.c_str(), tval)) {
 				wrong_answer("Expected float, got: %s", team.c_str());
 			}
-			if (!equal(tval, jval, float_abs_tol, float_rel_tol)) {
-				wrong_answer("Too large difference.\n Judge: %s\n Team: %s\n Difference: %Lg\n (abs tol %Lg rel tol %Lg)",
-							 judge.c_str(), team.c_str(), fabsl(jval-tval), float_abs_tol, float_rel_tol);
+			/* Finite values are compared with some tolerance */
```
Comment (vmcj) [inline sql/files/defaultdata/compare/compare.cc:156]: Is there a way how this can still stay another method? Moving this into the main function is not the right place IMO.

I do like the improvement for the rest though.

File: sql/files/defaultdata/compare/compare.cc [hunk 2]
```diff
@@ -176,16 +153,46 @@ int main(int argc, char **argv) {
 			if (!isfloat(team.c_str(), tval)) {
```
Comment (vmcj) [inline sql/files/defaultdata/compare/compare.cc:153]: I think you can also take this out as you have: https://github.com/DOMjudge/domjudge/pull/1863/files#diff-30594013bdc266f0b1b51ef87283423017bce078812a75f1c00a6a0716e1b7adR186

File: sql/files/defaultdata/compare/compare.cc [hunk 3]
```diff
@@ -72,33 +72,10 @@ FILE *openfeedback(const char *feedbackdir, const char *feedback, const char *wh
...
-	}
-	/* Values in different classes are always different. */
-	return 0;
-}
-
 const char *USAGE = "Usage: %s judge_in judge_ans feedback_dir [options] < team_out";
 
 int main(int argc, char **argv) {
-	if(argc < 4) {
+	if (argc < 4) {
```
Comment (eldering) [inline sql/files/defaultdata/compare/compare.cc:78]: Nitpick: could you make these unrelated layout changes in a separate commit?

**Review by eldering — COMMENTED** [review]
As a general comment: during the ICPC WFs we found a bug in this validator dealing with NULL bytes. I have a fix for that lying around that conflicts with this MR. Please don't merge this right away, so I can see what is the best way forward.

File: sql/files/defaultdata/compare/compare.cc (hunk 1 above)
Comment (mpsijm) [reply sql/files/defaultdata/compare/compare.cc]: Fair point! I've rewritten the function to not return an `int` but to directly invoke `wrong_answer` with the correct message :slightly_smiling_face:

File: sql/files/defaultdata/compare/compare.cc (hunk 2 above)
Comment (mpsijm) [reply sql/files/defaultdata/compare/compare.cc:193]: This `isfloat` is needed to populate `tval`, which is of type `flt`. Without it, the `isfinite`/`isinf`/`isnan` functions don't have a value for `tval` :slightly_smiling_face: (this should be clearer now that I've re-extracted the `compare_float` function)

File: sql/files/defaultdata/compare/compare.cc (hunk 3 above)
Comment (mpsijm) [reply sql/files/defaultdata/compare/compare.cc:118]: Sure, done! :smile:

File: sql/files/defaultdata/compare/compare.cc (hunk 2 above)
Comment (vmcj) [reply sql/files/defaultdata/compare/compare.cc:193]: Check!

**Comment by vmcj** [comment]
@mpsijm I'll set this to a draft, not because I think its not ready yet but to make sure we don't merge by accident waiting for @eldering to take a look at it.

**Comment by vmcj** [comment]
@mpsijm I'll close this one as #1917 is where @eldering improves further on this. If you disagree feel free to re-open.
